### PR TITLE
Add Vision keyboard

### PR DESCRIPTION
**Pull request by satt99**
## Description

Adds support for the Vision keyboard and its PCB.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Adds support for the Vision

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/satt/vision/bootloader_defs.h [hunk 1]
```diff
@@ -0,0 +1,7 @@
+/* Address for jumping to bootloader on STM32 chips. */
```
Comment (zvecr) [inline keyboards/satt/vision/bootloader_defs.h:1]: This file should be removed

File: keyboards/satt/vision/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        update_tri_layer(_LOWER, _RAISE, _ADJUST);
+      } else {
+        layer_off(_RAISE);
+        update_tri_layer(_LOWER, _RAISE, _ADJUST);
+      }
+      return false;
+      break;
+  }
+  return true;
+}
```
Comment (zvecr) [inline keyboards/satt/vision/keymaps/default/keymap.c:93]:
```suggestion
layer_state_t layer_state_set_user(layer_state_t state) {
  return update_tri_layer_state(state, _LOWER, _RAISE, _ADJUST);
}
```

File: keyboards/satt/vision/readme.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# Vision
+
+Vision Keyboard
+
+Keyboard Maintainer: [SatT](https://github.com/satt99)  
+Hardware Supported: STM32F072CBT6
```
Comment (zvecr) [inline keyboards/satt/vision/readme.md:6]:
```suggestion
* Keyboard Maintainer: [SatT](https://github.com/satt99)
* Hardware Supported: STM32F072CBT6
```

File: keyboards/satt/vision/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    _LOWER,
+    _RAISE,
+    _ADJUST
+};
+
+enum custom_keycodes {
+  LOWER = SAFE_RANGE,
+  RAISE,
+};
+
```
Comment (zvecr) [inline keyboards/satt/vision/keymaps/default/keymap.c:35]:
```suggestion
#define LOWER MO(_LOWER)
#define RAISE MO(_RAISE)
```

File: keyboards/satt/vision/config.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x5354
```
Comment (zvecr) [inline keyboards/satt/vision/config.h:21]: This ID is already used within `keyboards/unikorn/config.h`

File: keyboards/satt/vision/keymaps/default/keymap.c (hunk 2 above)
Comment (satt99) [reply keyboards/satt/vision/keymaps/default/keymap.c:93]: For the default layout I have an `MO(_ADJUST)` on my `_QWERTY` layer, so I need to use `update_tri_layer`, or else I wont be able to access `_ADJUST` with it. 

https://github.com/qmk/qmk_firmware/blob/master/docs/ref_functions.md#update_tri_layerx-y-z

File: keyboards/satt/vision/bootloader_defs.h (hunk 1 above)
Comment (satt99) [reply keyboards/satt/vision/bootloader_defs.h]: Removed 361e9c3

File: keyboards/satt/vision/config.h (hunk 5 above)
Comment (satt99) [reply keyboards/satt/vision/config.h]: Sorry, I overlooked it. Changed VID to `0x5454` 
1f9ded8

File: keyboards/satt/vision/config.h [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+#pragma once
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x5454
+#define PRODUCT_ID      0x5649
+#define DEVICE_VER      0x0001
+#define MANUFACTURER SatT
+#define PRODUCT Vision
+#define DESCRIPTION Vision Keyboard
```
Comment (fauxpark) [inline keyboards/satt/vision/config.h:26]:
```suggestion
```

File: keyboards/satt/vision/rules.mk [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+COMMAND_ENABLE = yes    # Commands for debug and configuration
+SLEEP_LED_ENABLE = no   # Breathing sleep LED during USB suspend
+NKRO_ENABLE = yes	    # USB Nkey Rollover
+CUSTOM_MATRIX = no      # Custom matrix file
+BACKLIGHT_ENABLE = no   # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no    # Enable keyboard RGB underglow
+MIDI_ENABLE = no        # MIDI support
+BLUETOOTH_ENABLE = no   # Enable Bluetooth
+AUDIO_ENABLE = no       # Audio output
+FAUXCLICKY_ENABLE = no  # Use buzzer to emulate clicky switches
```
Comment (fauxpark) [inline keyboards/satt/vision/rules.mk:20]:
```suggestion
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = yes        # Console for debug
COMMAND_ENABLE = yes        # Commands for debug and configuration
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

File: keyboards/satt/vision/keymaps/default/keymap.c (hunk 2 above)
Comment (zvecr) [reply keyboards/satt/vision/keymaps/default/keymap.c:93]: Missed that, lets ignore this suggestion then.

**Comment by satt99** [comment]
Hi, sorry to bug you guys.
Is there anything more I should do before the merge of this?

File: keyboards/satt/vision/info.json [hunk 8]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (fauxpark) [inline keyboards/satt/vision/info.json:1]: info.json renders like this in Configurator:
image

Comment (satt99) [reply keyboards/satt/vision/info.json:1]: Sorry, I forgot to check this. Fixed in 613aab3.
Although I had rotation set to 0 for all keys, the rotation center was non-zero (13.5, 3.5) which resulted in an error when converting XY-coordinates from KLE-raw with "Convert KLE raw to QMK info.json".

File: keyboards/satt/vision/vision.h [hunk 9]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include "quantum.h"
+
+#define LAYOUT_default( \
```
Comment (noroadsleft) [inline keyboards/satt/vision/vision.h:5]: As there only seems to be one layout possible, this should be named `LAYOUT`.

```suggestion
#define LAYOUT( \
```

File: keyboards/satt/vision/info.json [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+    "keyboard_name": "Vision",
+    "url": "",
+    "maintainer": "SatT",
+    "width": 16,
+    "height": 4,
+    "layouts": {
+        "LAYOUT_default": {
```
Comment (noroadsleft) [inline keyboards/satt/vision/info.json:8]:
```suggestion
        "LAYOUT": {
```

File: keyboards/satt/vision/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+#define ADJUST MO(_ADJUST)
+
+const uint16_t keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+	[_QWERTY] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/default/keymap.c:40]:
```suggestion
	[_QWERTY] = LAYOUT(
```

File: keyboards/satt/vision/keymaps/default/keymap.c [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        KC_PGUP, KC_ESC,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_BSPC, KC_DEL,
+        KC_PGDN, KC_TAB,  KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN,     KC_QUOT,
+                 KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, ADJUST,
+                 KC_LGUI,          KC_LALT,     LOWER,        KC_SPC,      RAISE,        KC_ENT,      KC_RALT,               KC_RCTL
+    ),
+
+	[_LOWER] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/default/keymap.c:47]:
```suggestion
	[_LOWER] = LAYOUT(
```

File: keyboards/satt/vision/keymaps/default/keymap.c [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        _______, _______, KC_EXLM, KC_AT,   KC_HASH, KC_DLR,  KC_PERC, KC_CIRC, KC_AMPR, KC_ASTR, KC_LPRN, KC_RPRN, _______, _______,
+        _______, _______, _______, _______, _______, _______, KC_GRV,  KC_BSLS, KC_MINS, KC_EQL,  KC_LBRC, KC_RBRC,      _______,
+                 _______, _______, _______, _______, _______, KC_TILD, _______, KC_PIPE, KC_UNDS, KC_PLUS, KC_LCBR, KC_RCBR, _______,
+                 _______,          _______,     _______,      _______,     _______,      _______, _______,                   _______
+    ),
+
+	[_RAISE] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/default/keymap.c:54]:
```suggestion
	[_RAISE] = LAYOUT(
```

File: keyboards/satt/vision/keymaps/default/keymap.c [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        _______, _______, KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, KC_LEFT, KC_DOWN, KC_UP,   KC_RGHT, KC_END,       _______,
+                 _______, _______, _______, _______, _______, _______, KC_HOME, XXXXXXX, KC_PGDN, KC_PGUP, XXXXXXX, _______, _______,
+                 _______,          _______,     _______,      _______,     _______,      _______, _______,                   _______
+    ),
+
+	[_ADJUST] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/default/keymap.c:61]:
```suggestion
	[_ADJUST] = LAYOUT(
```

File: keyboards/satt/vision/keymaps/via/keymap.c [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    _ADJUST
+};
+
+#define LOWER MO(_LOWER)
+#define RAISE MO(_RAISE)
+#define ADJUST MO(_ADJUST)
+
+const uint16_t keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+	[_QWERTY] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/via/keymap.c:37]:
```suggestion
	[_QWERTY] = LAYOUT(
```

File: keyboards/satt/vision/keymaps/via/keymap.c [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+        KC_PGUP, KC_ESC,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_BSPC, KC_DEL,
+        KC_PGDN, KC_TAB,  KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN,     KC_QUOT,
+                 KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, ADJUST,
+                 KC_LGUI,          KC_LALT,     LOWER,        KC_SPC,      RAISE,        KC_ENT,      KC_RALT,               KC_RCTL
+    ),
+
+	[_LOWER] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/via/keymap.c:44]:
```suggestion
	[_LOWER] = LAYOUT(
```

File: keyboards/satt/vision/keymaps/via/keymap.c [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+        _______, _______, KC_EXLM, KC_AT,   KC_HASH, KC_DLR,  KC_PERC, KC_CIRC, KC_AMPR, KC_ASTR, KC_LPRN, KC_RPRN, _______, _______,
+        _______, _______, _______, _______, _______, _______, KC_GRV,  KC_BSLS, KC_MINS, KC_EQL,  KC_LBRC, KC_RBRC,      _______,
+                 _______, _______, _______, _______, _______, KC_TILD, _______, KC_PIPE, KC_UNDS, KC_PLUS, KC_LCBR, KC_RCBR, _______,
+                 _______,          _______,     _______,      _______,     _______,      _______, _______,                   _______
+    ),
+
+	[_RAISE] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/via/keymap.c:51]:
```suggestion
	[_RAISE] = LAYOUT(
```

File: keyboards/satt/vision/keymaps/via/keymap.c [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+        _______, _______, KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, KC_LEFT, KC_DOWN, KC_UP,   KC_RGHT, KC_END,       _______,
+                 _______, _______, _______, _______, _______, _______, KC_HOME, XXXXXXX, KC_PGDN, KC_PGUP, XXXXXXX, _______, _______,
+                 _______,          _______,     _______,      _______,     _______,      _______, _______,                   _______
+    ),
+
+	[_ADJUST] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/via/keymap.c:58]:
```suggestion
	[_ADJUST] = LAYOUT(
```

File: keyboards/satt/vision/vision.h [hunk 19]
```diff
@@ -0,0 +1,15 @@
+#pragma once
```
Comment (noroadsleft) [inline keyboards/satt/vision/vision.h:1]: Need a license header here.

File: keyboards/satt/vision/vision.c [hunk 20]
```diff
@@ -0,0 +1 @@
+#include "vision.h"
```
Comment (noroadsleft) [inline keyboards/satt/vision/vision.c:1]: Need a license header here.

File: keyboards/satt/vision/vision.h (hunk 19 above)
Comment (satt99) [reply keyboards/satt/vision/vision.h:17]: Fixed in 31a59ea

File: keyboards/satt/vision/vision.c (hunk 20 above)
Comment (satt99) [reply keyboards/satt/vision/vision.c:17]: Fixed in 31a59ea

File: keyboards/satt/vision/keymaps/satt/keymap.c [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+#define IMOF_AL ALT_T(KC_F14)
+#define CTALDEL LCA(KC_DEL)
+#define ESC_ADJ LT(_ADJUST, KC_ESC)
+#define GUI_E LGUI(KC_E)
+#define GUI_R LGUI(KC_R)
+#define CTLGRV LCTL(KC_GRV)
+
+const uint16_t keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+	[_QWERTY] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/satt/keymap.c:60]:
```suggestion
	[_QWERTY] = LAYOUT(
```

File: keyboards/satt/vision/keymaps/satt/keymap.c [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+        GUI_E,   ESC_ADJ, KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_DEL, KC_DEL,
+        GUI_R,   TB_CTSF, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN,     KC_BSPC,
+                 KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    CTLGRV,  KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, KC_QUOT,
+                 KC_LGUI,          IMOF_AL,     LOWER,        SPC_CTL,     RAISE,        ENT_SFT, IMON_AL,                   KC_RCTL
+    ),
+
+	[_LOWER] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/satt/keymap.c:67]:
```suggestion
	[_LOWER] = LAYOUT(
```

File: keyboards/satt/vision/keymaps/satt/keymap.c [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+        _______, _______, KC_EXLM, KC_AT,   KC_HASH, KC_DLR,  KC_PERC, KC_CIRC, KC_AMPR, KC_ASTR, KC_LPRN, KC_RPRN, _______, _______,
+        _______, _______, _______, KC_F2,   KC_F5,   KC_F10,  KC_GRV,  KC_BSLS, KC_MINS, KC_EQL,  KC_LBRC, KC_RBRC,     _______,
+                 _______, _______, _______, _______, _______, KC_TILD, _______, KC_PIPE, KC_UNDS, KC_PLUS, KC_LCBR, KC_RCBR, _______,
+                 _______,          _______,     _______,      _______,     _______,      _______, _______,                   _______
+    ),
+
+	[_RAISE] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/satt/keymap.c:74]:
```suggestion
	[_RAISE] = LAYOUT(
```

File: keyboards/satt/vision/keymaps/satt/keymap.c [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+        _______, _______, KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, KC_LEFT, KC_DOWN, KC_UP,   KC_RGHT, KC_END,      _______,
+                 _______, _______, _______, _______, _______, _______, KC_HOME, XXXXXXX, KC_PGDN, KC_PGUP, XXXXXXX, _______, _______,
+                 _______,          _______,     _______,      _______,     _______,      _______, _______,                   _______
+    ),
+
+	[_PSEUDO_US] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/satt/keymap.c:81]:
```suggestion
	[_PSEUDO_US] = LAYOUT(
```

File: keyboards/satt/vision/keymaps/satt/keymap.c [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+        _______, _______, KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    _______, _______,
+        _______, _______, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    JIS2US,      _______,
+                 _______, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    JP_ZHTG, KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, JIS2US,
+                 _______,          _______,     P_LOWER,      _______,     P_RAISE,      _______, _______,                   _______
+    ),
+
+	[_PSEUDO_US_LOWER] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/satt/keymap.c:88]:
```suggestion
	[_PSEUDO_US_LOWER] = LAYOUT(
```

File: keyboards/satt/vision/keymaps/satt/keymap.c [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+        _______, _______, JP_EXLM, JP_AT,   JP_HASH, JP_DLR,  JP_PERC, JP_CIRC, JP_AMPR, JP_ASTR, JP_LPRN, JP_RPRN, _______, _______,
+        _______, _______, _______, KC_F2,   KC_F5,   KC_F10,  JP_GRV,  JP_BSLS, JP_MINS, JP_EQL,  JP_LBRC, JP_RBRC,     _______,
+                 _______, _______, _______, _______, _______, JP_TILD, _______, JP_PIPE, JP_UNDS, JP_PLUS, JP_LCBR, JP_RCBR, _______,
+                 _______,          _______,     _______,      _______,     _______,      _______, _______,                   _______
+    ),
+
+	[_PSEUDO_US_RAISE] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/satt/keymap.c:95]:
```suggestion
	[_PSEUDO_US_RAISE] = LAYOUT(
```

File: keyboards/satt/vision/keymaps/satt/keymap.c [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+        _______, _______, KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, KC_LEFT, KC_DOWN, KC_UP,   KC_RGHT, KC_END,      _______,
+                 _______, _______, _______, _______, _______, _______, KC_HOME, XXXXXXX, KC_PGDN, KC_PGUP, XXXXXXX, _______, _______,
+                 _______,          _______,     _______,      _______,     _______,      _______, _______,                   _______
+    ),
+
+	[_ADJUST] = LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/satt/vision/keymaps/satt/keymap.c:102]:
```suggestion
	[_ADJUST] = LAYOUT(
```

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**